### PR TITLE
Add reusable GitHub workflows "lint & tests" and "tag & publish" for JS projects

### PR DESCRIPTION
**Pull request by MrChocolatine**
### Add a reusable GitHub workflow "lint & tests" for Ember.js add-ons

This reusable workflow runs the typical jobs that Ember.js add-ons should run (as already implemented in [ember-cli-embedded]) to lint and test their codebase.

**It accepts 3 inputs from the caller workflow[^1]:**

- `ember_try_scenarios` (optional)
  A list of test scenarios to run with `ember-try`,
  If not specified, a predefined list of scenarios will be used.

  This can be used, for instance, to force `ember-try` not to run Embroider scenarios when it is not yet installed in the add-on. ([see an example])

- `nodejs_version` (optional)
  The version of Node.js to use.
  If not specified, the version 12 will be used ([which is still a LTS for the moment])

- `package_manager` (optional)
  Which package manager to use between `npm` and `yarn`.
  If not specified, `yarn` will be used.

One information, the step `Lint lockfile` is a custom addition in order to spot eventual malicious attacks (see https://github.com/lirantal/lockfile-lint).

[ember-cli-embedded]: https://github.com/peopledoc/ember-cli-embedded/blob/master/.github/workflows/ci.yml
[which is still a LTS for the moment]: https://nodejs.org/en/about/releases
[see an example]: https://github.com/peopledoc/ember-feature-controls/blob/79f4a86265a6fb96098f3f01856393e90e5f8fc3/.github/workflows/ci.yml#L18-L24

### Add a reusable GitHub workflow "tag & publish" for JS projects

When this new workflow can run it does:
- (1) Create a new `git tag` if a new version is detected in `package.json`
- (2-a) Create a new GitHub release linked to the newly created `git tag`
- (2-b) Publish the add-on to `npm` using the newly created `git tag`

_Note: the jobs `2-a` and `2-b` run in parallel._

**It accepts 2 inputs from the caller workflow[^1]:**

- `nodejs_version` (optional)
  The version of Node.js to use.
  If not specified, the version 12 will be used ([which is still a LTS for the moment])

- `package_manager` (optional)
  Which package manager to use between `npm` and `yarn`.
  If not specified, `yarn` will be used.

**It accepts 1 secret from the caller workflow[^1]:**

- `npm_automation_token` (required)
  An NPM automation token authorised to publish versions.

---

### Currently implemented in the repositories

- https://github.com/peopledoc/ember-feature-controls/pull/160
- https://github.com/peopledoc/ember-cli-embedded/pull/168
- https://github.com/peopledoc/eslint-config-peopledoc/pull/92
- https://github.com/peopledoc/ember-reading-time/pull/13
- https://github.com/peopledoc/ember-slugify/pull/149

[^1]: GitHub terminology
      See https://docs.github.com/en/actions/learn-github-actions/reusing-workflows

### CODE REVIEW

File: .github/workflows/emberjs-addons-tag-release-publish.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    timeout-minutes: 20
+    steps:
+      - uses: ncipollo/release-action@v1
+        with:
+          body: "\
+            Changelog: \
+            ${{ github.server_url }}/${{ github.repository }}\
+            /compare/\
+            v${{ needs.create_git_tag.outputs.old_version }}...v${{ needs.create_git_tag.outputs.new_version }}\
+          "
```
Comment (MrChocolatine) [inline .github/workflows/emberjs-addons-tag-release-publish.yml:57]: Why this syntax? => In order to split a long string over multiple lines, without getting a whitespace between each line.

See:
https://stackoverflow.com/questions/6268391/is-there-a-way-to-represent-a-long-string-that-doesnt-have-any-whitespace-on-mul

In comparison to:

```yaml
body: >
  Changelog: 
  my_repo
  /compare/
  v1...v2
```

Which would result in the following string: (notice the whitespaces between each section)

```
Changelog: my_repo /compare/ v1...v2
```

File: .github/workflows/emberjs-addons-tag-release-publish.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# GitHub Actions documentation:
+# https://docs.github.com/en/actions
+
+name: Create new `git tag`, create new GitHub release and publish to NPM
```
Comment (GreatWizard) [inline .github/workflows/emberjs-addons-tag-release-publish.yml:4]: Should we rename is `js-tag-release-publish`
If I'm correct we don't have any EmberJS related code in this workflow?

Comment (MrChocolatine) [reply .github/workflows/emberjs-addons-tag-release-publish.yml:4]: Ah true true, we can do that indeed, that would require an update to the CODEOWNERS file.

Following your suggestion, what about this @GreatWizard :

- Move the workflow `.github/workflows/emberjs-addons-ci.yml`
  to `.github/workflows/js/emberjs-addons-ci.yml`

- Move the workflow `.github/workflows/emberjs-addons-tag-release-publish.yml`
  to `.github/workflows/js/tag-release-publish.yml`

- Update the CODEOWNERS to make the Tribe owner of the pattern:

```
/.github/workflows/js/                  @peopledoc/tribe-js
```

File: .github/workflows/emberjs-addons-ci.yml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    runs-on: ubuntu-latest
+    timeout-minutes: 20
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v2
+        with:
+          cache: yarn
```
Comment (GreatWizard) [inline .github/workflows/emberjs-addons-ci.yml:45]: We should be possible to override this option.
Probabbly use npm by default and let the integration specify "yarn"

Comment (MrChocolatine) [reply .github/workflows/emberjs-addons-ci.yml:45]: We can indeed, do we want it in this first version? As all our repo use `yarn`.

Comment (GreatWizard) [reply .github/workflows/emberjs-addons-ci.yml:45]: right , I'm fine with it tbh :) at least add it in inputs with yarn by default ? just to be future proof

File: .github/workflows/emberjs-addons-tag-release-publish.yml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      - name: Set up Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version: ${{ inputs.nodejs_version }}
+          # Required if we want to authenticate during the `npm publish`:
+          # https://github.com/actions/setup-node/blob/v2.4.1/action.yml#L15-L16
+          registry-url: https://registry.npmjs.org
+
+      - name: Install Dependencies
+        run: yarn install --frozen-lockfile
```
Comment (GreatWizard) [inline .github/workflows/emberjs-addons-tag-release-publish.yml:80]: this command should be yarn/npm agnostic too?
In addition, we don't even specify the package manager for the cache

Comment (MrChocolatine) [reply .github/workflows/emberjs-addons-tag-release-publish.yml:80]: Again, as decided together, it is a first version that only targets our public add-ons on npm, which all use `yarn`. I did not feel the need to make the workflows more complex than they really need 🤔 .

File: .github/workflows/emberjs-addons-tag-release-publish.yml (hunk 2 above)
Comment (MrChocolatine) [reply .github/workflows/emberjs-addons-tag-release-publish.yml]: We can't put workflows into subdirectories è_é.

See commit b4c7f08d0cfe062afadeeb1c9b8a947f256296d6 .

**Review by GreatWizard — APPROVED** [review]

**Pull request merged**